src/libraries/System.Private.CoreLib/src/System/IO/FileStatus.Unix.cs

```diff
@@ -431,20 +488,24 @@ internal void RefreshCaches(ReadOnlySpan<char> path)
             // Check if the main path is a directory, or a link to a directory.
             int fileType = _fileCache.Mode & Interop.Sys.FileTypes.S_IFMT;
             bool isDirectory = fileType == Interop.Sys.FileTypes.S_IFDIR ||
-                               (fileType == Interop.Sys.FileTypes.S_IFLNK
-                                && Interop.Sys.Stat(path, out Interop.Sys.FileStatus target) == 0
-                                && (target.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR);
+                               (handle is null && // Don't follow links for SafeHandle APIs.
+                                fileType == Interop.Sys.FileTypes.S_IFLNK &&
```

Comment:
```suggestion
                               (fileType == Interop.Sys.FileTypes.S_IFLNK &&
```
In retrospect, this was overoptimizing.

For `handle is not null`, `fileType == Interop.Sys.FileTypes.S_IFLNK` will fail.